nemo/collections/asr/parts/k2/ctc.py

```diff
@@ -0,0 +1,115 @@
+
+import torch
+
+import k2
+from nemo.collections.asr.parts.k2.utils import get_tot_objf_and_num_frames
+
+
+class CTCLoss(torch.nn.Module):
+    """
+    Connectionist Temporal Classification (CTC) loss.
+    Ported from https://github.com/k2-fsa/snowfall/blob/master/snowfall/objectives/ctc.py
+    """
+    def __init__(
+            self,
+            num_classes: int,
+            blank: int,
+            reduction: str = 'mean_batch',
+            graph_type: str = 'topo',
+            L_inv: Optional[k2.Fsa] = None,
+            phones: Optional[k2.SymbolTable] = None,
+            words: Optional[k2.SymbolTable] = None,
+            oov: str = '<UNK>'
+    ):
+        super().__init__()
+        self.blank = blank
+        self.num_classes = num_classes
+        self.reduction = reduction
+        if graph_type == 'topo':
+            from nemo.collections.asr.parts.k2.graph_compilers import CtcTrainingTopologyCompiler as compiler
+            self.graph_compiler = compiler(self.num_classes)
+        elif graph_type == 'graph':
+            from nemo.collections.asr.parts.k2.graph_compilers import CtcTrainingGraphCompiler as compiler
+            # self.graph_compiler = compiler(L_inv, phones, words, oov)
+            raise NotImplementedError("Not adapted yet")
+        else:
+            raise ValueError(f"Invalid value of `graph_type`: {graph_type}.")
+
+    def create_supervision(self, input_lengths):
+        supervisions = torch.stack(
+            (
+                torch.tensor(range(input_lengths.shape[0])),
+                torch.zeros(input_lengths.shape[0]),
+                input_lengths.cpu(),
+            ),
+            1,
+        ).to(torch.int32)
+        # the duration column has to be sorted in decreasing order
+        order = torch.argsort(supervisions[:, -1], descending=True)
+        return supervisions, order
+
+    def forward(
+            self,
+            log_probs: torch.Tensor,
+            targets: torch.Tensor,
+            input_lengths: torch.Tensor,
+            target_lengths: torch.Tensor
+    ) -> torch.Tensor:
+        if self.blank != 0:
+            # rearrange log_probs to put blank at the first place
+            index = list(range(self.num_classes))
+            del index[self.blank]
+            index = torch.tensor([self.blank] + index).to(log_probs.device)
+            log_probs = torch.index_select(log_probs, -1, index)
+            # shift targets to emulate blank = 0
+            targets += 1
+        supervisions, order = self.create_supervision(input_lengths)
+        supervisions = supervisions[order]
+        log_probs = log_probs[order]
+        targets = targets[order]
+        target_lengths = target_lengths[order]
+        num_graphs = self.graph_compiler.compile(targets, target_lengths).to(log_probs.device)
+        dense_fsa_vec = k2.DenseFsaVec(log_probs, supervisions)
+
+        num_lats = k2.intersect_dense(num_graphs, dense_fsa_vec, 10.0)
```

Comment:
@GNroy You may want to try setting 10.0 to something like 1000.0. It's the equivalent of the "lattice beam" parameter in kaldi, which vanilla CTC does not use at all. This may cause the divergence in loss curves you told me about.